### PR TITLE
Adds valid height check for DEM intersection

### PR DESCRIPTION
**Pull request by jlaura**
## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

**Review by acpaquette — APPROVED** [review]

**Comment by jlaura** [comment]
@acpaquette  I just realized that this repo has not testing CI 😨!! This PR is passing locally (with warnings related to the PVL library). How do you want to proceed? I don't want this to become an uber PR. I can open another PR with testing setup, we can get that merged, and then re-review/re-run this one. Or you could pull this PR local, also verify that it is passing and then we can get an issue open. 

I did not mean for this to become a thing!

File: tests/test_csm.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+from plio.io.io_gdal import GeoDataset
+
+import csmapi
+from knoten import csm
+
+@pytest.fixture
+def mock_dem():
+    mock_dem = mock.MagicMock(spec_set=GeoDataset)
+    #mock_dem.read_array.return_value = 100
+    #mock_dem.latlon_to_pixel.return_value = (0.5,0.5)
```
Comment (acpaquette) [inline tests/test_csm.py]: Can these lines be removed?

File: tests/test_csm.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    mock_sensor.imageToGround.assert_called_once_with(pt, height)
+
+@mock.patch.object(csm, 'get_radii', return_value=(10,10))
+@mock.patch('pyproj.transformer.Transformer.transform', return_value=(0,0,0))
+@mock.patch.object(csm, '_compute_intersection_distance', return_value=0)
+def test_generate_ground_point_with_dtm(mock_dem, mock_sensor, pt):
+    # Passing the mock_dem fixture fails for some reason. The
+    # isinstance(obj, GeoDataset) check fails, causing the singldispath
+    # to never dispatch to the func under test.
+    mock_dem = mock.MagicMock(spec_set=GeoDataset)
```
Comment (acpaquette) [inline tests/test_csm.py]: Given the comment, should the `mock_dem` even be passed through here? Isn't the `mock_dem` variable that is defined shadowing the one that was passed?

File: tests/test_csm.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+from collections import namedtuple
+
+@mock.patch.object(csm, 'get_radii', return_value=(10,10))
+@mock.patch('pyproj.transformer.Transformer.transform', return_value=(0,0,0))
+def test_generate_ground_point_with_dtm_ndv(mock_dem, mock_sensor, pt):
+    # Passing the mock_dem fixture fails for some reason. The
+    # isinstance(obj, GeoDataset) check fails, causing the singldispath
+    # to never dispatch to the func under test.
+    mock_dem = mock.MagicMock(spec_set=GeoDataset)
```
Comment (acpaquette) [inline tests/test_csm.py]: See previous comment

**Review by acpaquette — COMMENTED** [review]
Couple comments to address, otherwise PR looks good!

File: tests/test_csm.py (hunk 1 above)
Comment (jlaura) [reply tests/test_csm.py]: Yes, I removed the whole mock_dem fixture since it isn't working

File: tests/test_csm.py (hunk 2 above)
Comment (jlaura) [reply tests/test_csm.py]: Yup, unused. Removed it!

File: tests/test_csm.py (hunk 3 above)
Comment (jlaura) [reply tests/test_csm.py]: Good catch - fixed.

File: tests/test_csm.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+def test_generate_ground_point_with_imagecoord(mock_sensor, pt):
+    height = 0.0
+    csm.generate_ground_point(height, pt, mock_sensor)
+    mock_sensor.imageToGround.assert_called_once_with(pt, height)
+
+@mock.patch.object(csm, 'get_radii', return_value=(10,10))
+@mock.patch('pyproj.transformer.Transformer.transform', return_value=(0,0,0))
+@mock.patch.object(csm, '_compute_intersection_distance', return_value=0)
+def test_generate_ground_point_with_dtm(mock_sensor, pt, _):
```
Comment (jlaura) [inline tests/test_csm.py]: As far as I can tell, pytest is counting the number of decorators and then assuming that n-decorators == n-args to the func. To get the test parser to work, I added the `_` arg. This seems super janky. Any thoughts on whether this approach should be different?

Comment (acpaquette) [reply tests/test_csm.py]: I'm not sure here, you might try putting the other `mock.patch.object` above the `mock.patch`. Some of pythons test suite is very order dependent

Comment (jlaura) [reply tests/test_csm.py]: This comment was 💰. The mock.patch.object needs to be passed into the test as an arg. The order is also important where the mocked stuff needs to be first in the test signature, then any fixtures. 

Local results with this setup: `5 passed, 4 warnings in 1.89s`

**Review by acpaquette — APPROVED** [review]

**Pull request merged**